Fixed cron job

OmarShehata opened the pull request:
Turns out the reason the crontab wasn't working was because of Windows line endings. I added a test to make sure it has the right endings. 

I also added `nano` in the Dockerfile just because it's useful when ssh'd and debugging.

The other big change is instead of checking if `cesium-concierge` has commented before (which will always be true) it checks if the last comment is by `cesium-concierge` to determine whether to post the initial or secondary stale PR message (so it fixes https://github.com/AnalyticalGraphicsInc/cesium-concierge/issues/77 )

Finally, it now logs every time it runs the stale PR check, which goes into a `cron.log` file, so it'll be easier in the future to track if anything has broken.

cesium-concierge commented:
@OmarShehata, thanks for the pull request! Maintainers, we have a signed CLA from @OmarShehata, so you can review this at any time.

I noticed that [CHANGES.md](https://github.com/AnalyticalGraphicsInc/cesium-concierge/blob/master/CHANGES.md) has not been updated. If this change updates the public API in any way, fixes a bug, or makes any non-trivial update, please add a bullet point to `CHANGES.md` and comment on this pull request so we know it was updated. For more info, see the [Pull Request Guidelines]( https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTING.md#pull-request-guidelines).

---

:sparkles: :sparkles: :sparkles:__I'm the Cesium Concierge and I'm here to test things.__ :sparkles: :sparkles: :sparkles:

ggetz commented:
Since this will be bumping a lot of PRs, we may want to consider https://github.com/AnalyticalGraphicsInc/cesium-concierge/issues/108 as part of this PR. 

Maybe the concierge can check to see if one of the comments in this thread is "Mute concierge 3 months" or something similar, and not bump until 3 months after the comment date.

ggetz commented:
> it checks if the last comment is by cesium-concierge to determine whether to post the initial or secondary stale PR message (so it fixes #77 )

Instead of checking the last comment, should we be checking the whole thread for an instance of the first message (since we do have access to what it was)? I do like this approach better than what we have before (as like you said, it fixes #77), so if that would take too much effort, I'm good with what you have here.

ggetz commented:
>> Instead of checking the last comment, should we be checking the whole thread for an instance of the first message (since we do have access to what it was)? I do like this approach better than what we have before (as like you said, it fixes #77), so if that would take too much effort, I'm good with what you have here.
>
>This is intentional. I think if Concierge does a 2nd bump and says "It hasn't been updated since I last commented, maybe this should be closed", but then activity starts on the PR again, and then dies down, it should say "Looks like this hasn't been updated in X days" as opposed to "It hasn't been updated since I last commented, maybe it should be closed". Like the current wording suggests the second message only comes up when the first warning is ignored.

But that's not exactly the logic you have here. You're checking if the last comment is the concierge.  For instance, if the concierge bumps and issue, then a maintainer asks a question that remains unanswered, the concierge will bump with the first message again. Is this the behavior we want? 

Do we even need two separate comments? I think we're not getting a lot of value out of it.

ggetz commented:
> If we really want it I think something as simple as "@concierge stop", the existence of that comment stops any concierge posting on that PR/issue, and deleting that comment would be how you undo that.

I like that!

>I think we should consider how often we'll be using this (for example, in this case, once this is deployed, it will bump a lot of posts that haven't been bumped in a while, which is the intended function, but if that turns out to be annoying, would someone actually go in and say "mute" on all those threads after the fact? Would that make the notifications any less annoying?)

I think it will be useful enough that it's worth it, for instance, see https://github.com/AnalyticalGraphicsInc/cesium/pull/6450 which just got bumped. The initial bump was good to get that conversation restarted, but we may want to pause until we have more information.

OmarShehata commented:
> But that's not exactly the logic you have here. You're checking if the last comment is the concierge. For instance, if the concierge bumps and issue, then a maintainer asks a question that remains unanswered, the concierge will bump with the first message again. Is this the behavior we want?

That's a good point, I hadn't considered that. I don't think we want that in that case.

> Do we even need two separate comments? I think we're not getting a lot of value out of it.

I don't think it's necessary. I think the point of the bump is to alert maintainers of something that hasn't been touched and should be cleaned somehow. The question is, should it keep bumping after that (unless explicitly told to stop) ? 

>  I think it will be useful enough that it's worth it, for instance, see AnalyticalGraphicsInc/cesium#6450 which just got bumped. The initial bump was good to get that conversation restarted, but we may want to pause until we have more information.

I could get that done real quick so that these can get pushed together! And maybe update the wording on the bump with instructions that say "to freeze concierge say 'stop' " or something to that effect.

ggetz commented:
> I think the point of the bump is to alert maintainers of something that hasn't been touched and should be cleaned somehow. The question is, should it keep bumping after that (unless explicitly told to stop) ?

I think it should keep bumping. 

So the consensus seems to be the following, correct?

1) Eliminate the two separate bumps in favor of one unified bump.
2) Keep bumping with that message every X days unless the stop command is detected.
3) Attach (minimal) instructions for stopping the commands in the bump template.

OmarShehata commented:
@ggetz yes that sounds exactly right!

OmarShehata commented:
Ok, lots of changes here. 

* In https://github.com/AnalyticalGraphicsInc/cesium-concierge/pull/106/commits/551b39528624c96c1c252f720ba6abc45a5e6637 I refactored it to use just 1 bump message instead of 2. I also added in a check for "stop". Right now that check is defined in `stalePullRequest.js` but I think it will make sense to move that to a separate file once we have another script that uses it. 
* In https://github.com/AnalyticalGraphicsInc/cesium-concierge/pull/106/commits/090dceef2ca6597e385e3c6c4e818cc47bbb8579 I discovered that if a PR has more than 30 comments, the "last comment" will be the 30th comment, because the Github API will paginate the results. I've refactored it to account for that, and always get the actual last comment. 
  * This might be something to look out for in other parts of concierge that don't account for this.
  * I tested this out, and another thing I forgot about and fixed here was that, concierge's bump message actually says @cesium-concierge stop inside of it, so I had to modify the stop check to only stop if _someone else_ asked it to stop.

ggetz commented:
> I also added in a check for "stop". Right now that check is defined in stalePullRequest.js but I think it will make sense to move that to a separate file once we have another script that uses it.

I agree, please write up an issue so we don't forget, or maybe comment in #108, so we don't forget!

> This might be something to look out for in other parts of concierge that don't account for this.

Yes, I suspect this is the case with https://github.com/AnalyticalGraphicsInc/cesium-concierge/issues/70

OmarShehata commented:
I just wrote up an issue (https://github.com/AnalyticalGraphicsInc/cesium-concierge/issues/110) for that.

Looks like #70 is exactly this same issue, so I'll fix that and then this PR should be ready!

OmarShehata commented:
#70 should be fixed now. It turned out to be a little more tricky because if you request the last page (or you're on the last page), then the returned `link` information will be `null`. So I had to account for that in both the pages and comments requests, to account separately for when there's > 30 items and when there's < 30. 

Should be ready to merge and deploy!

ggetz commented on bin/stalePullRequest.js:
```diff
@@ -1,92 +1,131 @@
+        headers: repositorySettings.headers,
+        json: true,
+        resolveWithFullResponse: true
+    })
```
Maybe we should consider sorting the comment by last updated? https://developer.github.com/v3/pulls/comments/#list-comments-on-a-pull-request

ggetz commented on bin/stalePullRequest.js:
```diff
@@ -1,92 +1,131 @@
+            var linkData = parseLink(response.headers.link);
+            if (Cesium.defined(linkData)) {
+                commentsUrl = linkData.last.url;
+            }
```
It looks like we're always making another comment request even if there isn't another page.

ggetz commented on bin/stalePullRequest.js:
```diff
@@ -1,92 +1,131 @@
+};
+
+stalePullRequest.foundStopComment = function (commentsJsonResponse) {
+    for(var i = 0; i < commentsJsonResponse.length; i++){
```
spacing

ggetz commented on bin/stalePullRequest.js:
```diff
@@ -1,92 +1,131 @@
+    for(var i = 0; i < commentsJsonResponse.length; i++){
+        var comment = commentsJsonResponse[i].body.toLowerCase();
+        var userName = commentsJsonResponse[i].user.login;
+        if (userName !== 'cesium-concierge' && comment.indexOf('@cesium-concierge stop') !== -1) {
```
I think we should have this command be configurable, or at least pulled out to where it can easily be found and edited, just in case.

OmarShehata replied:
I'm planning on doing that as a separate PR so this doesn't get too big (see https://github.com/AnalyticalGraphicsInc/cesium-concierge/issues/110)

ggetz commented on lib/templates/stalePullRequest.hbs:
```diff
@@ -0,0 +1,7 @@
+ No one has commented on this pull request in {{ maxDaysSinceUpdate }} days. Maintainers, can you review, merge or close to keep things tidy?
+
+ _I'm going to re-bump this in {{ maxDaysSinceUpdate }} days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._
+
```
extra whitespace

ggetz commented on lib/templates/stalePullRequest.hbs:
```diff
@@ -0,0 +1,7 @@
+Bump! 
```
I think `Bump!` may not be the _most_ polite thing here. :smirk: 

May leave the `Thanks!`?

Also just a thought- Do we also want to make the username of the user who opened the PR available to the PR template? This way we can @ mentioned them specifically and ask them to take action (since ultimately it's the responsibility of the user who opened the PR to get it merged).

OmarShehata replied:
I just feel like the concierge constantly starting every comment with "thanks" makes it start to lose its meaning. 

I changed it though, and added a `userName` parameter!

ggetz commented on specs/bin/crontabSpec.js:
```diff
@@ -0,0 +1,14 @@
+
+var fs = require('fs');
+
+describe('crontab', function () {
```
I think we should add a file to the repo like this: https://github.com/AnalyticalGraphicsInc/agi-cesium-cloud/blob/master/.gitattributes

rather than have a test.

ggetz commented on specs/bin/stalePullRequestSpec.js:
```diff
@@ -117,38 +145,35 @@ describe('stalePullRequest', function () {
+            if (options.resolveWithFullResponse === true) {
+                return Promise.resolve({headers: {link: '<https://url?page=2>; rel="next",<https://url?page=3>; rel="last"'}});
+            }
+            else if (options.url === commentsUrl + '?page=3') {
```
should be formatted `} else if (...) {`

OmarShehata replied:
Ah, turns out there was multiple places where it was like that!

ggetz commented on bin/stalePullRequest.js:
```diff
@@ -1,92 +1,131 @@
+    return (Date.now() - date.getTime()) / msPerDay;
+};
+
+stalePullRequest.foundStopComment = function (commentsJsonResponse) {
```
I don't think this function is tested at all. Try running the coverage task.

OmarShehata replied:
Added a test for it! Coverage looks a lot better now.

OmarShehata commented:
Each commit should take care of an individual item in your review @ggetz. It should be ready now I think.

ggetz commented:
Looks great ! This is going to be a huge help, thanks for the effort here @OmarShehata!
